src/logging/timestamp_checkpoint_manager.cpp

```diff
@@ -0,0 +1,1084 @@
+                db_catalog->GetDatabaseName().c_str());
+    } catch (Exception &e) {
+      LOG_DEBUG("Create database storage object %d '%s'", db_oid,
+                db_catalog->GetDatabaseName().c_str());
+
+      // create database storage object
+      database = new storage::Database(db_oid);
+      // TODO: This should be deprecated, dbname should only exists in pg_db
+      database->setDBName(db_catalog->GetDatabaseName());
+      storage_manager->AddDatabaseToStorageManager(database);
+
+      // put database object into rw_object_set
+      txn->RecordCreate(db_oid, INVALID_OID, INVALID_OID);
+    }
+
+    // recover table storage objects
+    size_t table_size = metadata_buffer.ReadLong();
+    for (oid_t table_idx = 0; table_idx < table_size; table_idx++) {
+      oid_t table_oid = metadata_buffer.ReadInt();
+
+      auto table_catalog = db_catalog->GetTableObject(table_oid);
+      PELOTON_ASSERT(table_catalog != nullptr);
+
+      LOG_DEBUG("Create table object %d '%s'", table_oid,
+                table_catalog->GetTableName().c_str());
+
+      // recover column information
+      std::vector<catalog::Column> columns;
+      size_t column_count = metadata_buffer.ReadLong();
+      for (oid_t column_idx = 0; column_idx < column_count; column_idx++) {
+        oid_t column_oid = metadata_buffer.ReadInt();
+        size_t column_length = metadata_buffer.ReadLong();
+
+        auto column_catalog = table_catalog->GetColumnObject(column_oid);
+
+        // create column storage object
+        // ToDo: Column should be recovered from catalog
+        auto column = catalog::Column(
+            column_catalog->GetColumnType(), column_length,
+            column_catalog->GetColumnName(), column_catalog->IsInlined(),
+            column_catalog->GetColumnOffset());
+
+        // recover column constraints
+        // ToDo: Constraint should be recovered from catalog
+        size_t column_constraint_count = metadata_buffer.ReadLong();
+        for (oid_t constraint_idx = 0; constraint_idx < column_constraint_count;
+             constraint_idx++) {
+          auto column_constraint = catalog::Constraint::DeserializeFrom(
+              metadata_buffer, column.GetType());
+          // Foreign key constraint will be stored by DataTable deserializer
+          if (column_constraint.GetType() != ConstraintType::FOREIGN) {
+            column.AddConstraint(column_constraint);
+          }
+        }
+
+        // Set a column in the vector in order of the column_id in ColumnCatalogObject
+        // ToDo: just insert by push_back function
+        auto column_itr = columns.begin();
+        for (oid_t idx_count = START_OID; idx_count < column_oid; idx_count++) {
+        	if (column_itr == columns.end() ||
+        			column_itr->GetOffset() > column.GetOffset()) {
+        		break;
+        	} else {
+        		column_itr++;
+        	}
+        }
+      	columns.insert(column_itr, column);
```

Comment:
Are we skipping any specific columns? I'm wondering why we can't just insert by `push_back` now.

Reply:
Here is that it arranges the order of columns for column storage objects. Because column catalog objects obtained from column catalog are not ordered by column id. And the column storage object is managed column id by the vector iterator. This means there is no column id in column storage object. This is the reason why we doesn't use push_back. To use push_back, we need to implement the column id into column storage object.

Reply:
I simplify this insertion method by vector iterator using [].